Unification of Nephele Memory with PACT & Removal of Checkpointing and Compression

rmetzger opened the pull request:
This is issue https://github.com/dimalabs/ozone/issues/24

This pull request reworks the Buffer/FileBuffer/MemoryBuffer classes of Nephele. 
The FileBuffer has been removed as it is not really used. Only the checkpointing feature of nephele was using it. This feature has been completely removed with this pull request.

The MemoryBuffer now wraps a MemorySegment instead of a ByteBuffer in order to unify the different memory models.
Please note that the ByteBuffers were allocated using .allocateDirect(), that is the memory is from outside the heap space. I had to increase the heap space for the integration tests (maven failsafe) because the MemorySegment allocates memory on the heap (using byte-arrays)

I annotated the CoGroupConnectedComponentsITCase test with @Ignore because it causes the integration tests to fail.

rmetzger committed:
removed the FileBuffer from Nephele, and everything that uses it (Checkpointing, FileChannels)

rmetzger committed:
Changed Nephele MemoryBuffer to wrap PACT MemorySegment instead of java.nio.ByteBuffer

StephanEwen commented on nephele/nephele-common/src/main/java/eu/stratosphere/nephele/io/channels/Buffer.java:
```diff
@@ -180,4 +163,17 @@ public final void recycleBuffer()
+	 *         end of the stream.
+	 * @throws IOException Thrown if an error occurs while writing to the {@link WritableByteChannel} object.
+	 */
+	public abstract int read(WritableByteChannel writableByteChannel) throws IOException;
```
Can we remame this method into writeTo(...). The name is totally irritating.

StephanEwen commented on nephele/nephele-common/src/main/java/eu/stratosphere/nephele/io/channels/Buffer.java:
```diff
@@ -16,11 +16,11 @@
 package eu.stratosphere.nephele.io.channels;
 
 import java.io.IOException;
-import java.nio.ByteBuffer;
```
Is it possible to remove the logic that assumes the buffer is first in writable mode and then in readable mode? When we move to the buffer-oriented execution, all the buffer ever sees is a memory segment that carries. The contents of the memory segment if never changed while the buffer carries it.

StephanEwen replied:
What is the logic behind the duplicate method? Where would it duplicate to, when no source of memory is specified.

rmetzger replied:
The ByteBuffer-based implementation was doing the same thing. It results in two MemoryBuffers that are backed by the same MemorySegment.
This is only used by the transfer envelope (it looks like the Buffer is duplicated for each remote receiver)
The MemoryBufferRecycler counts the references for each MemoryBuffer.

StephanEwen commented on nephele/nephele-examples/src/main/java/eu/stratosphere/nephele/example/compression/CompressionTest.java:
```diff
@@ -1,88 +0,0 @@
-/***********************************************************************************************************************
```
Do we need to remove the compression test? If the reason is that they reference file channels, then please add them again and only change the file channels to memory channels.

If the reason is that they require nio buffers, then keep them removed.

StephanEwen commented on nephele/nephele-examples/src/main/java/eu/stratosphere/nephele/example/events/EventExample.java:
```diff
@@ -1,99 +0,0 @@
-/***********************************************************************************************************************
```
The event examples should not be removed. Where they reference the file buffers, please change the channel type to memory.

rmetzger replied:
I added the event examples again.
BUT it seems that most nephele-examples are somewhat useless because they use hard-coded paths like "file:///Users/casp/test2.txt", "file:///Users/casp/EventTask.jar" 
(the Grep, Event, Compression examples are affected).
Do you want me to remove them?

StephanEwen commented on nephele/nephele-examples/src/main/java/eu/stratosphere/nephele/example/events/EventReceiver.java:
```diff
@@ -1,52 +0,0 @@
-/***********************************************************************************************************************
```
The event examples should not be removed. Where they reference the file buffers, please change the channel type to memory.

StephanEwen commented on nephele/nephele-examples/src/main/java/eu/stratosphere/nephele/example/events/EventSender.java:
```diff
@@ -1,71 +0,0 @@
-/***********************************************************************************************************************
```
The event examples should not be removed. Where they reference the file buffers, please change the channel type to memory.

StephanEwen commented on nephele/nephele-examples/src/main/java/eu/stratosphere/nephele/example/events/MyEventListener.java:
```diff
@@ -1,49 +0,0 @@
-/***********************************************************************************************************************
```
The event examples should not be removed. Where they reference the file buffers, please change the channel type to memory.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/executiongraph/ExecutionGraph.java:
```diff
@@ -945,15 +926,9 @@ public void repairStages() {
-					}
+				if (stageNumber != succeedingNumber) {
+					LOG.error(edge.getSourceVertex() + " and " + edge.getTargetVertex()
+						+ " are assigned to different stages although not connected by a file channel");
```
It seems this case should through a genuine exception. There really are no different stages currently without file channels. However, we might want to introduce them in a different way based on the knowledge about whether a task materializes (sort, hashtable, pipeline braker).

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/channels/MemoryBuffer.java:
```diff
@@ -21,82 +21,84 @@
 	@Override
-	public int read(final ByteBuffer dst) throws IOException {
-
+	public int read(ByteBuffer dst) throws IOException {
```
Please check if this method is used at all, if not, please remove it. 
What are situations in which contents is copied to a byte buffer?

rmetzger replied:
Network I/O uses this method.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/channels/MemoryBuffer.java:
```diff
@@ -21,82 +21,84 @@
-			dst.put(this.byteBuffer);
+			
+		final int oldIndex = index;
+		for(; index < limit; index++) {
```
This will be rather inefficient, as it goes through the repeated position checks and translation in both the memory segment and the byte buffer. This should really be one bulk copy. We need to add respective methods to MemorySegment, if we keep this method.

StephanEwen replied:
An alternative to adding the method to the MemorySegment is that you can use wrap(offset, limit) to wrap the memory segment and then do a bulk write to the byte buffer.

StephanEwen replied:
Update: MemorySegment has a get(ByteBuffer) method, which bulk gets the array data into a byte buffer.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/channels/MemoryBuffer.java:
```diff
@@ -257,25 +257,58 @@ public void copyToBuffer(final Buffer destinationBuffer) throws IOException {
 		}
+		final MemoryBuffer target = (MemoryBuffer) destinationBuffer;
+		
+		System.arraycopy(this.getMemorySegment().getBackingArray(), 0,
```
Move this to a method in memory segment. The getBackingArray() method is now deprecated and may disappear.

StephanEwen replied:
Update: MemorySegment has copyTo(...) method, which should be used.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/channels/MemoryBuffer.java:
```diff
@@ -21,82 +21,84 @@
 	@Override
-	public int read(final WritableByteChannel writableByteChannel) throws IOException {
-
+	public int read(WritableByteChannel writableByteChannel) throws IOException {
```
should we rename this to writeTo(...)? I find the name "read()" confusing. At least "readInto(...)" would be more understandable.

uce replied:
The same confusing naming in SerializationBuffer. I renamed it to writeTo() there after coordinating with Robert.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/channels/MemoryBuffer.java:
```diff
@@ -257,25 +257,58 @@ public void copyToBuffer(final Buffer destinationBuffer) throws IOException {
-
-		destinationBuffer.finishWritePhase();
+		final int initialPos = position();
+		while(src.hasRemaining()) {
```
Use the bulk put method of memory segment "put(ByteBuffer)"

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/compression/AbstractCompressor.java:
```diff
@@ -16,10 +16,10 @@
 package eu.stratosphere.nephele.io.compression;
```
I assume the compressor / decompressor code will fail, because it is JNI code and expects to get a DirectByteBuffer.

In the long run, we could add an LZ4 (seems to be fastest) compressor implemented in java. The project on "https://github.com/jpountz/lz4-java" has a cool implementation that uses unsafe (like our memory buffers) to get an extremely fast compressor fully java based (no JNI and hence no need for directbytebuffers)

rmetzger replied:
Okay, I did not pay much attention to everything related to compression because I could not find any actual implementations of the AbstractDecompressor. Eclipse seems to ignore the nephele/nephele-compression-*/src files as source files. So I thought the Compression stuff was just an academic exercise. 

I'm going to fix this.

The provided "Test case" is named as such but is not one, it is just a main() method with hard coded paths. Thats another reason why I just killed much of the compression code. 
So the compression code was never tested. I'm not sure if somebody used it and if it is working!

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/io/compression/AbstractDecompressor.java:
```diff
@@ -16,13 +16,13 @@
 package eu.stratosphere.nephele.io.compression;
 
 import java.io.IOException;
-import java.nio.ByteBuffer;
```
Same as for the compressor holds for the decompressor.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/jobmanager/scheduler/RecoveryLogic.java:
```diff
@@ -27,7 +27,6 @@
 import org.apache.commons.logging.LogFactory;
 
 import eu.stratosphere.nephele.execution.ExecutionState;
-import eu.stratosphere.nephele.executiongraph.CheckpointState;
```
Can you make sure that recovery logic currently simply cancels the job and restarts it?

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/taskmanager/TaskCheckpointState.java:
```diff
@@ -19,11 +19,9 @@
 import java.io.DataOutput;
```
Can we fully remove this file?

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/taskmanager/bufferprovider/LocalBufferPool.java:
```diff
@@ -165,14 +156,10 @@ private Buffer requestBufferInternal(final int minimumSizeOfBuffer, final boolea
+					return BufferFactory.createFromMemory(minimumSizeOfBuffer, memSeg, this.bufferPoolConnector);
 				}
 			}
-
```
You removed the asynchronous event listener concept. Was that fully related to checkpointing? I somehow thought it had also to do with asynchronous rebalancing of the buffer pools. Please verify that it can really be removed, I am a little unsure here! Removing it even though it has a role beyond checkpointing will almost definitely lead to potential deadlocks.

rmetzger replied:
There are two implementations of the asynchronousEventOccurred() method. One in the ReplayTaskContext, which I removed, and another in the RuntimeTaskContext. The implementation there executes this code

```
        if (this.ephemeralCheckpoint == null) {
            return;
        }

        if (!this.ephemeralCheckpoint.isUndecided()) {
            return;
        }

        // TODO: Remove this and implement decision logic for ephemeral checkpoint
        LOG.error("Checkpoint state of " + this.task.getRuntimeEnvironment().getTaskNameWithIndex() + " is UNDECIDED");
        this.ephemeralCheckpoint.setCheckpointDecisionSynchronously(false);
```

Which seems to do nothing, especially not related to buffer rebalancing.

StephanEwen commented on nephele/nephele-server/src/main/java/eu/stratosphere/nephele/taskmanager/transferenvelope/SpillingQueue.java:
```diff
@@ -15,24 +15,16 @@
 
```
The spilling queue can probably go. It was used for asynchronous shipping and required file buffers.

StephanEwen commented on nephele/nephele-server/src/test/java/eu/stratosphere/nephele/executiongraph/ExecutionGraphTest.java:
```diff
@@ -783,136 +783,136 @@ public void testConvertJobGraphToExecutionGraph3() {
-			}
-		}
-	}
+//	/*
```
This used to be a pretty important test. It does not inherently depend on file buffers. Please re-enable it, replacing file buffers with memory buffers.

StephanEwen commented on nephele/nephele-server/src/test/java/eu/stratosphere/nephele/jobmanager/JobManagerITCase.java:
```diff
@@ -60,11 +60,6 @@
  */
```
Please re-enable this important test!

Remove the file buffer references, but do not disable the entire test, please.

StephanEwen commented on pact/pact-tests/src/test/java/eu/stratosphere/pact/test/iterative/CoGroupConnectedComponentsITCase.java:
```diff
@@ -28,6 +29,7 @@
 import eu.stratosphere.pact.test.iterative.nephele.ConnectedComponentsNepheleITCase;
 import eu.stratosphere.pact.test.util.TestBase2;
 
+@Ignore
```
Remove this! The test now works fine.

StephanEwen commented on pom.xml:
```diff
@@ -346,7 +346,7 @@
 				</executions>
 				<configuration>
-					<argLine>-Xms512m -Xmx512m</argLine>
+					<argLine>-Xms512m -Xmx1536m</argLine>
```
Have you experienced this high memory demand? It used to rune fine without.

the -Xms version should also be removed. Most tests don't need to fire up half a gig of mem in order to run.

rmetzger replied:
The old implementation was using ByteBuffer.allocateDirect(), hence the memory was not allocated on the heap. Since we now use MemorySegments which allocate their byteArrays on the heap, we have increased memory demands.

I had failing test cases without this change!

Commit:
Remove checkpointing and envelope consumption logic

Signed-off-by: Robert Metzger <metzgerr@web.de>

rmetzger committed:
Addressed Stephans comments
new memory segments methods

rmetzger committed:
Removed compression from nephele

rmetzger commented:
Okay, I addressed all comments and pushed again. Please note that the pull request contains a commit from Ufuk.

The commit that removes the compression feature from nephele is also contained in this pull request.
If you want, I can open a separate pull request (its a 5k lines change). 
I'm not sure anymore, if it is a good idea to remove the compression feature this way. It seems that it was never used and it does probably not work. But the code that I removed probably contains all the infrastructure to control compression.

There is one issue left: the testConvertJobGraphToExecutionGraph4() of ExecutionGraphTest.
The test case has causes an error (not a failure. JUnit continues testing after erroneous tests).
@StephanEwen: if you are able to fix the issue in a few minutes, I would be glad if you could fix it. If it is not trivial for you, let me know, than I'll invest more time.

dimalabs commented:
Manually merged and committed in b394c0b5e44e6c90df140eb18b9b37a2413190e5
